mlir/lib/Dialect/Rock/Transforms/LowerRockReduce.cpp

```diff
@@ -0,0 +1,212 @@
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ============================================================
+//
+// This pass converts rock.reduce into the TransformingFor loop
+// using global atomics.
+//
+//===-----------------------------------------------------===//
+#include "mlir/Dialect/Arithmetic/IR/Arithmetic.h"
+#include "mlir/Dialect/GPU/IR/GPUDialect.h"
+#include "mlir/Dialect/Rock/IR/Rock.h"
+#include "mlir/Dialect/Rock/IR/TransformMapBuilder.h"
+#include "mlir/Dialect/Rock/Passes.h"
+#include "mlir/Dialect/Rock/utility/loweringUtils.h"
+#include "mlir/Dialect/Rock/utility/transformMapUtils.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+#include "llvm/Support/Debug.h"
+#include <memory>
+
+namespace mlir {
+namespace rock {
+#define GEN_PASS_DEF_ROCKLOWERREDUCEPASS
+#include "mlir/Dialect/Rock/Passes.h.inc"
+} // namespace rock
+} // namespace mlir
+
+#define DEBUG_TYPE "rock-lower-rock-reduce"
+
+using namespace mlir;
+using namespace mlir::rock;
+
+namespace {
+class RockLowerReducePass
+    : public rock::impl::RockLowerReducePassBase<RockLowerReducePass> {
+  void runOnOperation() override;
+};
+
+struct ReduceRewritePattern : public OpConversionPattern<ReduceOp> {
+  using OpConversionPattern<ReduceOp>::OpConversionPattern;
+  LogicalResult
+  matchAndRewrite(ReduceOp op, ReduceOpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override;
+};
+} // end namespace
+
+// This function will create a view of the tensor that is being
+// reduced from arbitary dimension sizes to bid x tid x iter space.
+static ArrayAttr createThreadViewMaps(Value redInput, int64_t blockSize,
+                                      int64_t gridSize, Location loc,
+                                      PatternRewriter &rewriter) {
+  int64_t totalThreads = gridSize * blockSize;
+  ShapedType inpShape = redInput.getType().cast<ShapedType>();
+  int64_t elementCount = inpShape.getNumElements();
+  int64_t dataPerThread = (elementCount + (totalThreads - 1)) / totalThreads;
+
+  SmallVector<int64_t> topSizes;
+  int64_t topSizeProduct = 1;
+  SmallVector<unsigned int> topDims;
+  for (auto dimAndSize : llvm::enumerate(inpShape.getShape())) {
+    size_t dim = dimAndSize.index();
+    int64_t dimSize = dimAndSize.value();
+    topSizes.push_back(dimSize);
+    topSizeProduct *= dimSize;
+    topDims.push_back(dim);
+  }
+
+  BottomUpTMBuilder threadsToInpTensor(rewriter, topSizes, loc);
+  SmallVector<StringRef, 4> topNameRefs;
```

Comment:
For clarity, this should be `bottonNameRefs` - you've got the thing backwards. The names you're getting are the names of the output dimensions.

Reply:
done!